raiden/utils/upgrades.py

```diff
@@ -0,0 +1,87 @@
+import os
+import shutil
+import sqlite3
+from pathlib import Path
+
+import sqlitebck
+import structlog
+
+from raiden.exceptions import RaidenDBUpgradeError
+from raiden.storage.serialize import JSONSerializer
+from raiden.storage.sqlite import RAIDEN_DB_VERSION, SQLiteStorage
+from raiden.storage.versions import older_db_file
+from raiden.utils.migrations.v16_to_v17 import upgrade_initiator_manager
+
+UPGRADES_LIST = [
+    upgrade_initiator_manager,
+]
+
+
+log = structlog.get_logger(__name__)
+
+
+class UpgradeManager:
+    """ This class is responsible for figuring out which migrations
+    need to be executed in order to bring the database up to date
+    with the current implementation.
+    """
+    def __init__(self, db_filename: str):
+        self._current_db_filename = Path(db_filename)
+
+    def run(self):
+        """
+        The `_db_filename` is going to hold the filename of the database
+        with the new version. However, the previous version's data
+        is going to exist in a file whose name contains the old version.
+        Therefore, running the migration means that we have to copy
+        all data to the current version's database, execute the migration
+        functions.
+        """
+        if self._current_db_filename.exists():
```

Comment:
We have a problem here, checking the file name is not sufficient. If the program crashes after the copy of the database, the data is backup but the upgrade did not happen.

This needs to open the database and query for the version inside the database.

Reply:
I think the best is to delete the file if the version in the database is not the one we expected and there is a previous version, and then running the whole upgrade again.